crates/proto/src/dnssec/tsig.rs

```diff
@@ -26,11 +27,158 @@ use super::rdata::tsig::{
     TSIG, TsigAlgorithm, make_tsig_record, message_tbs, signed_bitmessage_to_buf,
 };
 use super::{DnsSecError, DnsSecErrorKind};
+use crate::dnssec::rdata::tsig::TsigError;
 use crate::error::{ProtoError, ProtoResult};
+use crate::op::message::ResponseSigner;
 use crate::op::{Message, MessageSignature, MessageSigner, MessageVerifier};
 use crate::rr::{Name, RData};
+use crate::serialize::binary::{BinEncoder, EncodeMode};
 use crate::xfer::DnsResponse;
 
+/// Context for a TSIG response, used to construct a TSIG response signer
+#[allow(missing_copy_implementations)]
+pub struct TSigResponseContext {
+    request_id: u16,
+    time: u64,
+}
+
+impl TSigResponseContext {
+    /// Create a new TSIG response context
+    pub fn new(request_id: u16, time: u64) -> Self {
+        Self { request_id, time }
+    }
+
+    /// Yield a response signer for a valid request signature
+    pub fn sign(
+        self,
+        req_sig: &TSIG,
+        error: Option<TsigError>,
```

Comment:
I guess we should document that callers should not pass `BadSig` or `BadKey` in here?